src/d/d_player_mdl.cpp

```diff
@@ -0,0 +1,1959 @@
+    {                   "at", 255, 255, 57, 57,  2},
+    {                 "take", 255, 255, 57, 57, 57},
+    {                "takeL", 255, 255, 57, 57, 57},
+    {                "takeR", 255, 255, 57, 57, 57},
+    {                "Take2", 255, 255, 57, 57, 57},
+    {             "waitAtoS",   9,   1, 57, 57,  2},
+    {                 "damD", 255,  56,  6, 57,  6},
+    {           "Ashootwait",  13,   1, 57, 57,  2},
+    {              "Arelord",   8,   3, 57, 57,  2},
+    {               "Ashoot",   8,   3, 57, 57,  2},
+    {          "ArelordTame",   8,   3, 57, 57,  2},
+    {           "Ashootwait",  13,   1, 57, 57,  2},
+    {              "Arelord",   8,   3, 57, 57,  2},
+    {               "Ashoot",   8,   3, 57, 57,  2},
+    {                "damSw",  13,   6,  6, 57,  6},
+    {               "HSwait",   8,   1, 57, 57,  2},
+    {           "VacuumWait",  12,   5, 57, 57, 57},
+    {          "MPPoseFront",  13, 255, 57, 57,  2},
+    {         "MPPoseAround",  13, 255, 57, 57, 57},
+    {     "MPPoseAroundBack",  13, 255, 57, 57,  2},
+    {      "MPPoseDashFront",  13, 255, 57, 57,  2},
+    {     "MPPoseDashAround",  13, 255, 57, 57, 57},
+    { "MPPoseDashAroundBack",  13, 255, 57, 57,  2},
+    {           "LiftB_UtoD",   7,   0, 57, 57,  2},
+    {           "LiftS_UtoD",   7,   0, 57, 57, 57},
+    {            "LiftLWalk",   7,   0, 57, 57,  2},
+    {          "MPGuardPose", 255, 255, 57, 57, 57},
+    {             "RockRide", 255,  56, 12, 57, 12},
+    {           "BinDrinkSt", 255,   5, 13, 13, 13},
+    {             "BinDrink", 255,   5, 14, 57, 14},
+    {           "BinDrinkEd", 255,   5, 15, 57, 15},
+    {             "HarpPose",   7,   6,  1,  0,  0},
+    {           "HarpPlayGo",   7,   1,  1,  0,  0},
+    {         "HarpPlayCome",   7,   0,  1,  0,  0},
+    {              "BinHold", 255,   5, 57, 57, 57},
+    {            "BinShield",   9,   5, 24, 24, 24},
+};
+#pragma pop
+
+const s32 daPlayerModelBase_c::sUnkDuration = 0x2400;
+
+const u8 daPlayerModelBase_c::sShieldDurabilities[10] = {16, 24, 32, 24, 36, 48, 12, 16, 20, 48};
+const u8 daPlayerModelBase_c::sShieldRegenTimes[10] = {0, 0, 0, 0, 0, 0, 90, 90, 90, 0};
+
+/**
+ * Watch the player's hand material indices and change the ResMat
+ * so that the right hand mats are visible and all others are hidden.
+ */
+void daPlBaseScnObjCallback_c::ExecCallback_DRAW_OPA(
+    nw4r::g3d::ScnObj::Timing timing, nw4r::g3d::ScnObj *pObj, u32 param, void *pInfo
+) {
+    if (timing != nw4r::g3d::ScnObj::CALLBACK_TIMING_C) {
+        nw4r::g3d::ScnMdl *pMdl = nw4r::g3d::ScnObj::DynamicCast<nw4r::g3d::ScnMdl>(pObj);
+        nw4r::g3d::ResMdl mdl = pMdl->GetResMdl();
+        nw4r::g3d::ResShp shp(nullptr);
+        for (int i = 0; i < 2; i++) {
+            u8 defaultMat = sSavedHandMats[i];
+            // TODO: Is there a way to avoid this cast?
+            u8 playerHandMat = static_cast<daPlayerModelBase_c *>(field_0x0C)->getHandMat(i);
```

Comment:
is there a reason the field isnt directly `daPlayerModelBase_c*`?

Reply:
This callback class is owned by `daPlayerModelBase_c`, so we'd need to forward declare `daPlayerModelBase_c`, but at the same time the player is passed as a constructor argument to `dScnCallback_c`, which needs a `dAcObjBase_c`. But with a forward declaration we cannot express that the player class is a subclass of object. So either this callback class needs to be moved into the player class namespace, which messes up vtable order, or a static cast has to happen *somewhere*.